File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
'''logic
    1. dfs를 이용한 최대최소를 구하는 방법'''
import sys

input = sys.stdin.readline
num = int(input())

num_list = list(map(int, input().split()))

# 연산자
op = list(map(int, input().split()))

# 최솟값 최댓값 초기화 !
max_num = -1e9
min_num = 1e9

# dfs 메소드 구현
def dfs(depth, total, add, sub, mul, div):
    global max_num, min_num
    if depth == num:
        max_num = max(total, max_num)
        min_num = min(total, min_num)
        return
    
    if add:
        dfs(depth +1, total + num_list[depth], add-1, sub, mul, div)
    if sub:
        dfs(depth+1, total - num_list[depth], add, sub-1, mul, div)
    if mul:
        dfs(depth+1, total * num_list[depth], add, sub, mul-1, div)
    if div:
        dfs(depth+1, int(total/num_list[depth]), add, sub, mul, div-1)
    
dfs(1, num_list[0], op[0], op[1], op[2], op[3])
print(max_num)
print(min_num)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int N, maxNum = Integer.MIN_VALUE, minNum = Integer.MAX_VALUE;
    static int[] numbers;
    static int[] count;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        numbers = new int[N];
        count = new int[4]; // 연산자 갯수 -> +, -, *, /

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            numbers[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            count[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, numbers[0], count[0], count[1], count[2], count[3]); // depth, total

        System.out.println(maxNum);
        System.out.println(minNum);
    }

    private static void dfs(int depth, int total, int plus, int minus, int multi, int div) {
        // 기저
        if (depth == N) {
            minNum = Math.min(minNum, total);
            maxNum = Math.max(maxNum, total);
            return;
        }

        // 유도
        if (plus > 0) {
            dfs(depth + 1, total + numbers[depth], plus - 1, minus, multi, div);
        }
        if (minus > 0) {
            dfs(depth + 1, total - numbers[depth], plus, minus - 1, multi, div);
        }
        if (multi > 0) {
            dfs(depth + 1, total * numbers[depth], plus, minus, multi - 1, div);
        }
        if (div > 0) {
            dfs(depth + 1, total / numbers[depth], plus, minus, multi, div - 1);
        }

    }

}
